File: golang/mtconfig/mtconfigserver.go
package mtconfig

import (
    "../zk"
    "io/ioutil"
    "net/http"
    "strings"
    "time"
)

type MtConfigServer struct {
    zklist []string
    cache_config map[string]map[string]string   //path-keys-values
    zookeeper *zk.Conn
}


func NewMtConfigServer(config_url string) (*MtConfigServer, error) {
    if(config_url == "") {
        config_url = "http://xxx.xxx.com/api/zkserverlist"
    }

    //try to get zklist by http get request
    res, err := http.Get(config_url)
    if(err != nil) {
        return nil, err
    }

    bodyByte, _ := ioutil.ReadAll(res.Body)

    res.Body.Close()

    ipList := strings.TrimSpace(string(bodyByte))
    ipArray := strings.Split(ipList, ",")

    zkCli, _, err := zk.Connect(ipArray, time.Second* 10)

    if(err != nil) {
        return nil, err
    }


    return &MtConfigServer{
                zklist: ipArray,
                zookeeper: zkCli,
                cache_config: make(map[string]map[string]string),
            }, err
}

func (p *MtConfigServer) Get(domain string, node string, key string) (string, error) {
    path := "/config/" + domain + "/" + node
    value, err := p.findInCache(path, key)

    if(value == "") {
        //find it in zk
        err = p.getFromZk(path)
        if(err != nil){
            return "", err
        }
    }else {
        return value, err
    }

    value, err = p.findInCache(path, key)

    return value, err
}

func (p *MtConfigServer) findInCache(path_ string, key_ string) (string, error) {
    //try to find config from cache_config
    for path, mapItem := range p.cache_config {
        if(path == path_){
            for key, value := range mapItem {
                if(key == key_){
                    return value, nil
                }
            }
        }
    }

    return "", nil
}

func (p *MtConfigServer) deleteFromCache(path string) {
    delete(p.cache_config, path)
}

func (p* MtConfigServer) getFromZk(path string) error {
    //first to delete the path cache
    p.deleteFromCache(path)

    p.zookeeper.SetCallBackFunc(p.onCallBack)
    //get data from zk
    data, _, _, err := p.zookeeper.GetW(path)
    
    if(err != nil) {
        return err
    }
    
    //parse the data //
    trimData := strings.TrimSpace(string(data))
    dataArry := strings.Split(trimData, "\n\n")
    
    item := make(map[string]string)

    for _, data := range dataArry {
        subDataArray := strings.Split(data, "=")
        
        item[strings.TrimSpace(string(subDataArray[0]))] = string(strings.TrimSpace(subDataArray[1]))
    }

    p.cache_config[path] = item

    return nil
}


func (p *MtConfigServer) onCallBack(path string) {
    p.deleteFromCache(path)
}

File: python/mtconfig/mtconfigserver.py
# coding: utf-8
# auth weishouyang@meituan.com 2014-06-30
# mtconfigserver

import sys
import zookeeper, time, threading, urllib

class mtconfigserver():
    def __init__(self, config_server_="http://xxx.xxx.com/"):
        self.config_server = config_server_
        self.cache_config = []  
        self.zk_path = "/config"
        self.zklist = self.getZkServer()
        zookeeper.set_debug_level(zookeeper.LOG_LEVEL_ERROR)
        self.zookeeper = zookeeper.init(self.zklist)

    def getZkServer(self):
        url = urllib.urlopen(self.config_server + "/api/zkserverlist")
        data = url.read()
        return data
    
    def get(self, domain, node, key):
        path = self.zk_path + "/" + domain + "/" + node
        value = self._find_in_cache(path, key)
        if value != "":
            return value

        self._get_from_zk(path)
        return self._find_in_cache(path, key)


    def _find_in_cache(self, path, key):
        for cs in self.cache_config:
            if cs.path==path:
                for ci in cs.items:
                    if ci.key==key:
                        return ci.value
        return ""

    def _delete_from_cache(self, path):
        i = 0
        while i<len(self.cache_config):
            if self.cache_config[i].path == path:
                del self.cache_config[i]
                break
            i=i+1 

    def _get_from_zk(self, path):
        self._delete_from_cache(path)
        content = zookeeper.get(self.zookeeper, path, self._onCallBack)
        self._parse(path, content[0])

    def _parse(self, path, config_data):
        strip_data = config_data.strip()
        data_array = strip_data.split('\n\n')
        cs = config_set(path)
        i = 0
        while i<len(data_array):
            item_array = data_array[i].split('=')
            if len(item_array)==2:
                key = item_array[0].strip()
                value = item_array[1].strip()
                ci = config_item(key, value)
                cs.items.append(ci)
            i=i+1

        self.cache_config.append(cs)


    def _onCallBack(self, handler, event_type, stat, path):
        self._get_from_zk(path)

class config_set():
    def __init__(self, path):
        self.path = path
        self.items = []

class config_item():
    def __init__(self, key, value):
        self.key = key
        self.value = value



